[SPARK-18186] Migrate HiveUDAFFunction to TypedImperativeAggregate for partial aggregation support

liancheng opened the pull request:
## What changes were proposed in this pull request?

While being evaluated in Spark SQL, Hive UDAFs don't support partial aggregation. This PR migrates `HiveUDAFFunction`s to `TypedImperativeAggregate`, which already provides partial aggregation support for aggregate functions that may use arbitrary Java objects as aggregation states.

The following snippet shows the effect of this PR:

```scala
import org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMax
sql(s"CREATE FUNCTION hive_max AS '${classOf[GenericUDAFMax].getName}'")

spark.range(100).createOrReplaceTempView("t")

// A query using both Spark SQL native `max` and Hive `max`
sql(s"SELECT max(id), hive_max(id) FROM t").explain()
```

Before this PR:

```
== Physical Plan ==
SortAggregate(key=[], functions=[max(id#1L), default.hive_max(default.hive_max, HiveFunctionWrapper(org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMax,org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMax@7475f57e), id#1L, false, 0, 0)])
+- Exchange SinglePartition
   +- *Range (0, 100, step=1, splits=Some(1))
```

After this PR:

```
== Physical Plan ==
SortAggregate(key=[], functions=[max(id#1L), default.hive_max(default.hive_max, HiveFunctionWrapper(org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMax,org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMax@5e18a6a7), id#1L, false, 0, 0)])
+- Exchange SinglePartition
   +- SortAggregate(key=[], functions=[partial_max(id#1L), partial_default.hive_max(default.hive_max, HiveFunctionWrapper(org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMax,org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMax@5e18a6a7), id#1L, false, 0, 0)])
      +- *Range (0, 100, step=1, splits=Some(1))
```

The tricky part of the PR is mostly about updating and passing around aggregation states of `HiveUDAFFunction`s since the aggregation state of a Hive UDAF may appear in three different forms. Let's take a look at the testing `MockUDAF` added in this PR as an example. This UDAF computes the count of non-null values together with the count of nulls of a given column. Its aggregation state may appear as the following forms at different time:

1. A `MockUDAFBuffer`, which is a concrete subclass of `GenericUDAFEvaluator.AggregationBuffer`

   The form used by Hive UDAF API. This form is required by the following scenarios:

   - Calling `GenericUDAFEvaluator.iterate()` to update an existing aggregation state with new input values.
   - Calling `GenericUDAFEvaluator.terminate()` to get the final aggregated value from an existing aggregation state.
   - Calling `GenericUDAFEvaluator.merge()` to merge other aggregation states into an existing aggregation state.

     The existing aggregation state to be updated must be in this form.

   Conversions:

   - To form 2:

     `GenericUDAFEvaluator.terminatePartial()`

   - To form 3:

     Convert to form 2 first, and then to 3.

2. An `Object[]` array containing two `java.lang.Long` values.

   The form used to interact with Hive's `ObjectInspector`s. This form is required by the following scenarios:

   - Calling `GenericUDAFEvaluator.terminatePartial()` to convert an existing aggregation state in form 1 to form 2.
   - Calling `GenericUDAFEvaluator.merge()` to merge other aggregation states into an existing aggregation state.

     The input aggregation state must be in this form.

   Conversions:

   - To form 1:

     No direct method. Have to create an empty `AggregationBuffer` and merge it into the empty buffer.

   - To form 3:

     `unwrapperFor()`/`unwrap()` method of `HiveInspectors`

3. The byte array that holds data of an `UnsafeRow` with two `LongType` fields.

   The form used by Spark SQL to shuffle partial aggregation results. This form is required because `TypedImperativeAggregate` always asks its subclasses to serialize their aggregation states into a byte array.

   Conversions:

   - To form 1:

     Convert to form 2 first, and then to 1.

   - To form 2:

     `wrapperFor()`/`wrap()` method of `HiveInspectors`

Here're some micro-benchmark results produced by the most recent master and this PR branch.

Master:

```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_92-b14 on Mac OS X 10.10.5
Intel(R) Core(TM) i7-4960HQ CPU @ 2.60GHz

hive udaf vs spark af:                   Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
w/o groupBy                                    339 /  372          3.1         323.2       1.0X
w/ groupBy                                     503 /  529          2.1         479.7       0.7X
```

This PR:

```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_92-b14 on Mac OS X 10.10.5
Intel(R) Core(TM) i7-4960HQ CPU @ 2.60GHz

hive udaf vs spark af:                   Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
w/o groupBy                                    116 /  126          9.0         110.8       1.0X
w/ groupBy                                     151 /  159          6.9         144.0       0.8X
```

Benchmark code snippet:

```scala
  test("Hive UDAF benchmark") {
    val N = 1 << 20

    sparkSession.sql(s"CREATE TEMPORARY FUNCTION hive_max AS '${classOf[GenericUDAFMax].getName}'")

    val benchmark = new Benchmark(
      name = "hive udaf vs spark af",
      valuesPerIteration = N,
      minNumIters = 5,
      warmupTime = 5.seconds,
      minTime = 5.seconds,
      outputPerIteration = true
    )

    benchmark.addCase("w/o groupBy") { _ =>
      sparkSession.range(N).agg("id" -> "hive_max").collect()
    }

    benchmark.addCase("w/ groupBy") { _ =>
      sparkSession.range(N).groupBy($"id" % 10).agg("id" -> "hive_max").collect()
    }

    benchmark.run()

    sparkSession.sql(s"DROP TEMPORARY FUNCTION IF EXISTS hive_max")
  }
```

## How was this patch tested?

New test suite `HiveUDAFSuite` is added.

liancheng committed:
Initial draft

liancheng committed:
Fix test failures

liancheng commented:
Will add more details in the PR description soon.

SparkQA commented:
**[Test build #67842 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67842/consoleFull)** for PR 15703 at commit [`c0029f1`](https://github.com/apache/spark/commit/c0029f1a529935c263f9c83691cf84921b343e67).

liancheng committed:
Comment update

SparkQA commented:
**[Test build #67844 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67844/consoleFull)** for PR 15703 at commit [`5a23a97`](https://github.com/apache/spark/commit/5a23a979c5e6a61f847b146a1cb656418054d955).

SparkQA commented:
**[Test build #67842 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67842/consoleFull)** for PR 15703 at commit [`c0029f1`](https://github.com/apache/spark/commit/c0029f1a529935c263f9c83691cf84921b343e67).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67842/
Test FAILed.

SparkQA commented:
**[Test build #67844 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67844/consoleFull)** for PR 15703 at commit [`5a23a97`](https://github.com/apache/spark/commit/5a23a979c5e6a61f847b146a1cb656418054d955).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67844/
Test FAILed.

rxin commented:
cc @tejasapatil

liancheng commented:
I can't reproduce those test failures when executing failed test cases individually. Seems that it's related to execution order. Still investigating.

tejasapatil commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -293,69 +307,57 @@ private[hive] case class HiveUDAFFunction(
+  private lazy val partialResultInspector =
+    function.init(GenericUDAFEvaluator.Mode.PARTIAL1, inspectors)
+
+  // The following two lines initializes `function: GenericUDAFEvaluator` eagerly. These two fields
```
This feels little weird. If you want eager init and not serialize, why not remove `lazy` and just keep it `@transient` ?

tejasapatil commented:
This will surely improve performance for the UDAFs where the data shrinks (eg. `max` as you pointed out). I am not sure if it would be better for UDAFs like `GenericUDAFCollectSet`, `GenericUDAFCollectList` where the aggregation does not shrink the data (it might be worse because of conversion cost ?).

It would be good to see some perf numbers to compare (native Spark UDFs) vs (before this change) vs (after the change).

liancheng commented:
@tejasapatil For `collect_set` and `collect_list`, we'll simply migrate them to `TypedImperativeAggregate` and so that they become Spark native aggregate functions. We can also handle other built-in Hive UDAFs that share the same issue in this approach.

In the long run, this PR is mostly useful for running legacy UDAFs developed by the users rather than Hive built-in ones since we'd prefer to migrate all the built-in ones to Spark at last.

liancheng commented:
I found that I'm handling bridged UDAFs properly, which caused a few test failures. Working on it.

liancheng committed:
Remove the @transient lazy val hack

SparkQA commented:
**[Test build #67910 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67910/consoleFull)** for PR 15703 at commit [`88ef7ea`](https://github.com/apache/spark/commit/88ef7ea701f694c9ce1c2f9ffd7ef0b9b90bb202).

liancheng commented:
@tejasapatil Another point that I'd like to add is that even if the performance for a single UDAF like `GenericUDAFCollectList` regresses, you still have performance gains if such UDAFs are used together with other aggregate functions within a single query. Before this PR, all the aggregate functions have to be evaluated without partial aggregation support in this case.

tejasapatil commented:
There is no doubt that this PR is not going to make things better. I'm already sold :)

SparkQA commented:
**[Test build #67910 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67910/consoleFull)** for PR 15703 at commit [`88ef7ea`](https://github.com/apache/spark/commit/88ef7ea701f694c9ce1c2f9ffd7ef0b9b90bb202).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67910/
Test FAILed.

liancheng committed:
Properly initialize Hive UDAF evaluators

Hive UDAFs are sensitive to aggregation mode, and must be initialized
with proper modes before being used. Basically, it means that you can't
use an evaluator initialized with mode PARTIAL1 to merge two aggregation
states (although it still works for aggregate functions whose partial
result type is the same as the final result type).

liancheng commented:
It turned out that I didn't initialize Hive UDAF evaluators properly.

Quoted from commit message of my previous commit:

> Hive UDAFs are sensitive to aggregation mode, and must be initialized with proper modes before being used. Basically, it means that you can't use an evaluator initialized with mode `PARTIAL1` to merge two aggregation states (although it still works for aggregate functions whose partial result type is the same as the final result type).

SparkQA commented:
**[Test build #67927 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67927/consoleFull)** for PR 15703 at commit [`3af4f21`](https://github.com/apache/spark/commit/3af4f211ceaf749568fd9fb819c25c5d64892dac).

liancheng committed:
Update comments

SparkQA commented:
**[Test build #67930 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67930/consoleFull)** for PR 15703 at commit [`6f28688`](https://github.com/apache/spark/commit/6f28688fd5b3ed1fd02a1082c121812cecbfe001).

SparkQA commented:
**[Test build #67927 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67927/consoleFull)** for PR 15703 at commit [`3af4f21`](https://github.com/apache/spark/commit/3af4f211ceaf749568fd9fb819c25c5d64892dac).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67927/
Test FAILed.

SparkQA commented:
**[Test build #67930 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67930/consoleFull)** for PR 15703 at commit [`6f28688`](https://github.com/apache/spark/commit/6f28688fd5b3ed1fd02a1082c121812cecbfe001).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67930/
Test FAILed.

liancheng committed:
Should create testing UDAFs as temporary functions and drop them while cleaning up

SparkQA commented:
**[Test build #67938 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67938/consoleFull)** for PR 15703 at commit [`1a2f4c7`](https://github.com/apache/spark/commit/1a2f4c7e32cde26b611d5af8b4197ce5afbe93d9).

SparkQA commented:
**[Test build #67938 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67938/consoleFull)** for PR 15703 at commit [`1a2f4c7`](https://github.com/apache/spark/commit/1a2f4c7e32cde26b611d5af8b4197ce5afbe93d9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67938/
Test PASSed.

liancheng commented:
OK, now it's ready for review and merge.

cc @yhuai @JoshRosen @cloud-fan

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -289,73 +302,77 @@ private[hive] case class HiveUDAFFunction(
-  override def eval(input: InternalRow): Any = unwrapper(function.evaluate(buffer))
+  private lazy val finalModeEvaluator = {
+    val parameterInfo = new SimpleGenericUDAFParameterInfo(inputInspectors, false, false)
+    resolver.getEvaluator(parameterInfo)
```
These 2 lines are duplicated many times, should we abstract them to a method?

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -365,4 +382,66 @@ private[hive] case class HiveUDAFFunction(
+
+    private val partialResultWrapper = wrapperFor(partialResultInspector, partialResultDataType)
+
+    private val projection = UnsafeProjection.create(Array(partialResultDataType))
```
Can you try to run hive udaf in spark shell? IIUC, we can't create unsafe projection inside UDAF

liancheng replied:
It does work as expected:

```
scala> sql("CREATE TEMPORARY FUNCTION hive_max AS 'org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMax'")
res0: org.apache.spark.sql.DataFrame = []

scala> spark.range(100).createOrReplaceTempView("t")

scala> sql("SELECT hive_max(id) FROM t").explain()
== Physical Plan ==
SortAggregate(key=[], functions=[hive_max(hive_max, HiveFunctionWrapper(org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMax,org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMax@144792d5), id#1L, false, 0, 0)])
+- Exchange SinglePartition
   +- SortAggregate(key=[], functions=[partial_hive_max(hive_max, HiveFunctionWrapper(org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMax,org.apache.hadoop.hive.ql.udf.generic.GenericUDAFMax@144792d5), id#1L, false, 0, 0)])
      +- *Range (0, 100, step=1, splits=Some(8))

scala> sql("SELECT hive_max(id) FROM t").show()
+-------------+
|hive_max( id)|
+-------------+
|           99|
+-------------+
```

Why do you think `UnsafeProjection` can't be used here?

cloud-fan replied:
I tried to use unsafe projection in `percentile_approx` before, but failed in spark shell, maybe it's a different problem, nvm.

liancheng committed:
Minor refactoring

SparkQA commented:
**[Test build #68433 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68433/consoleFull)** for PR 15703 at commit [`4c6284b`](https://github.com/apache/spark/commit/4c6284bac2ee9d98304e07fca3c6f3805f942eb8).

SparkQA commented:
**[Test build #68433 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68433/consoleFull)** for PR 15703 at commit [`4c6284b`](https://github.com/apache/spark/commit/4c6284bac2ee9d98304e07fca3c6f3805f942eb8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68433/
Test FAILed.

liancheng commented:
The last build failure was because of a logical conflict between this PR and the master branch. Resolving it.

SparkQA commented:
**[Test build #68488 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68488/consoleFull)** for PR 15703 at commit [`3ffafe7`](https://github.com/apache/spark/commit/3ffafe7947757142b99f36ad4d81b471e5188d3b).

SparkQA commented:
**[Test build #68488 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68488/consoleFull)** for PR 15703 at commit [`3ffafe7`](https://github.com/apache/spark/commit/3ffafe7947757142b99f36ad4d81b471e5188d3b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68488/
Test FAILed.

SparkQA commented:
**[Test build #68491 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68491/consoleFull)** for PR 15703 at commit [`b418cd7`](https://github.com/apache/spark/commit/b418cd75d1edf7e1f48d3112d7c42d6f95985878).

liancheng committed:
Fix test failures

We're now using ObjectHashAggregateExec instead of SortAggregateExec to
evaluate Hive UDAFs.

SparkQA commented:
**[Test build #68493 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68493/consoleFull)** for PR 15703 at commit [`e88db5c`](https://github.com/apache/spark/commit/e88db5c9638678592f0f8836097e88d47558ffc1).

SparkQA commented:
**[Test build #68491 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68491/consoleFull)** for PR 15703 at commit [`b418cd7`](https://github.com/apache/spark/commit/b418cd75d1edf7e1f48d3112d7c42d6f95985878).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68491/
Test FAILed.

SparkQA commented:
**[Test build #68493 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68493/consoleFull)** for PR 15703 at commit [`e88db5c`](https://github.com/apache/spark/commit/e88db5c9638678592f0f8836097e88d47558ffc1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68493/
Test PASSed.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -289,73 +302,75 @@ private[hive] case class HiveUDAFFunction(
+  // The UDAF evaluator used to consume raw input rows and produce partial aggregation results.
   @transient
-  private lazy val function = functionAndInspector._1
+  private lazy val partial1ModeEvaluator = newEvaluator()
```
Do we need to make it a lazy val since partialResultInspector is uses it right below?

yhuai replied:
Should we always call init to make the code consistent?

liancheng replied:
Yes. It has to be transient and lazy so that it's also available on executor side since Hive UDAF evaluators are not serializable.

yhuai replied:
ok. I think in general we should avoid of using this pattern. If we have to use it now, let's explain it in the comment.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -289,73 +302,75 @@ private[hive] case class HiveUDAFFunction(
   override def nullable: Boolean = true
 
-  override def supportsPartial: Boolean = false
+  override def supportsPartial: Boolean = true
```
Is there any Hive UDAF that does not support partial aggregation?

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -365,4 +380,66 @@ private[hive] case class HiveUDAFFunction(
+    partial1ModeEvaluator.getNewAggregationBuffer
+
+  @transient
+  private lazy val inputProjection = new InterpretedProjection(children)
```
Why `InterpretedProjection`?

liancheng replied:
Copied from the original code. Replaced with `UnsafeProjection` instead. Thanks.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -365,4 +380,66 @@ private[hive] case class HiveUDAFFunction(
+
+  override def deserialize(bytes: Array[Byte]): AggregationBuffer = {
+    aggBufferSerDe.deserialize(bytes)
+  }
```
Let's add docs to explain what these functions are doing.

liancheng replied:
Done.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -365,4 +380,66 @@ private[hive] case class HiveUDAFFunction(
+  }
+
+  override def merge(buffer: AggregationBuffer, input: AggregationBuffer): Unit = {
+    partial2ModeEvaluator.merge(buffer, partial1ModeEvaluator.terminatePartial(input))
```
Let's explain what we are trying to do using `partial1ModeEvaluator.terminatePartial(input)`.

liancheng replied:
Comment added.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -365,4 +380,66 @@ private[hive] case class HiveUDAFFunction(
+
+    private val mutableRow = new GenericInternalRow(1)
+
+    def serialize(buffer: AggregationBuffer): Array[Byte] = {
```
doc?

liancheng replied:
Done.

yhuai commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDAFSuite.scala:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  test("customized Hive UDAF") {
+    val df = sql("SELECT mock(value) FROM t GROUP BY key % 2")
```
How about we also keep the key?

yhuai commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDAFSuite.scala:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  test("built-in Hive UDAF") {
+    val df = sql("SELECT hive_max(key) FROM t GROUP BY key % 2")
```
Also keep the key?

liancheng replied:
Done. Thanks.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -263,8 +265,19 @@ private[hive] case class HiveGenericUDTF(
+ *  1. a Spark SQL value, or
+ *  2. an instance of some concrete `GenericUDAFEvaluator.AggregationBuffer` class, or
+ *  3. a Java object that can be inspected using the `ObjectInspector` returned by the
+ *     `GenericUDAFEvaluator.init()` method.
```
Besides of explaining what are these three formats, let's also explain when we will use each of them.

yhuai replied:
(we can just put the pr description to here)

yhuai replied:
(is the doc below enough?)

liancheng replied:
Thanks, added.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -289,73 +302,75 @@ private[hive] case class HiveUDAFFunction(
+  // Hive `ObjectInspector`s for all child expressions (input parameters of the function).
   @transient
-  private lazy val inspectors = children.map(toInspector).toArray
+  private lazy val inputInspectors = children.map(toInspector).toArray
```
Let's add docs to explain when these internal vals are used (like which vals are needed for a given mode).

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -289,73 +302,75 @@ private[hive] case class HiveUDAFFunction(
-  private lazy val function = functionAndInspector._1
+  private lazy val partial1ModeEvaluator = newEvaluator()
 
+  // Hive `ObjectInspector` used to inspect partial aggregation results.
```
Partial aggregation result is aggregation buffer, right?

liancheng replied:
Yea. They are those objects returned by `terminatePartial()`, which is the inspectable version of Hive `AggregationBuffer`.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -289,73 +302,75 @@ private[hive] case class HiveUDAFFunction(
-  // in the superclass because that will lead to initialization ordering issues.
-  override val inputAggBufferAttributes: Seq[AttributeReference] = Nil
+  @transient
+  private lazy val aggBufferSerDe: AggregationBufferSerDe = new AggregationBufferSerDe
```
doc

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -365,4 +380,66 @@ private[hive] case class HiveUDAFFunction(
+      val unsafeRow = projection(mutableRow)
+      val bytes = ByteBuffer.allocate(unsafeRow.getSizeInBytes)
+      unsafeRow.writeTo(bytes)
+      bytes.array()
```
Should we just use `unsafeRow.getBytes`?

liancheng replied:
Aren't they equivalent in this case? `UnsafeRow.getBytes` also performs some more checks that are not necessary here.

cloud-fan replied:
but you also create an unnecessary `ByteBuffer`... as they are equivalent, doesn't `unsafeRow.getBytes` simpler?

cloud-fan replied:
Actually they are different. If the buffer type is fixed length, then the `unsafeRow` is just a fixed-length bytes array, and `UnsafeRow.getBytes` will just return that array, instead of copying the memory.

liancheng committed:
Address PR comments

SparkQA commented:
**[Test build #68734 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68734/consoleFull)** for PR 15703 at commit [`ca3978c`](https://github.com/apache/spark/commit/ca3978c7f368225d349d91477ca13aa2d5b9a3fa).

liancheng commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -365,4 +380,66 @@ private[hive] case class HiveUDAFFunction(
+  }
+
+  // Helper class used to de/serialize Hive UDAF `AggregationBuffer` objects
+  private class AggregationBufferSerDe {
```
We can, but it doesn't seem to be necessary. Make it a nested class also simplifies implementation since it has access to fields of the outer class.

SparkQA commented:
**[Test build #68734 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68734/consoleFull)** for PR 15703 at commit [`ca3978c`](https://github.com/apache/spark/commit/ca3978c7f368225d349d91477ca13aa2d5b9a3fa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68734/
Test PASSed.

yhuai commented:
Code changes looks good to me. Let's also do a benchmark to sanity check our implementation.

yhuai commented:
LGTM. Merging to master.

liancheng commented:
Thanks everyone for the review!

pgandhi999 commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:
```diff
@@ -365,4 +394,74 @@ private[hive] case class HiveUDAFFunction(
+    // buffer in the 3rd format mentioned in the ScalaDoc of this class. Originally, Hive converts
+    // this `AggregationBuffer`s into this format before shuffling partial aggregation results, and
+    // calls `GenericUDAFEvaluator.terminatePartial()` to do the conversion.
+    partial2ModeEvaluator.merge(buffer, partial1ModeEvaluator.terminatePartial(input))
```
If we follow the code flow from interfaces.scala, we see that the results of aggregation buffer mode in PARTIAL2 is merged with the aggregation buffer in PARTIAL1. I am new to Spark and Hive, so just wanted to know the reason behind the above behaviour. If there are any docs suggesting this, do let me know. Thank you.
